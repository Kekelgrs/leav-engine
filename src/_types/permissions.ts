export enum PermissionTypes {
    RECORD = 'record'
}

export enum RecordPermissions {
    ACCESS = 'access',
    CREATE = 'create',
    EDIT = 'edit',
    DELETE = 'delete'
}

export enum PermissionsRelations {
    AND = 'and',
    OR = 'or'
}

export interface ITreePermissionsConf {
    permissionTreeAttributes: [string];
    relation: PermissionsRelations;
}

export interface IPermissionsTreeTarget {
    tree: string;
    library: string;
    id: string | number;
}

export interface IPermission {
    type: PermissionTypes;
    usersGroup: string;
    actions: {[name: string]: boolean | null};
    permissionTreeTarget?: IPermissionsTreeTarget | string;
}
