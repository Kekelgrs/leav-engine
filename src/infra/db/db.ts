import {Database} from 'arangojs';

interface IDeps {
    config?: any;
}

const _getDbConnection = (config: any): Database => {
    return new Database({
        url: config.db.url
    });
};

export default function(deps: IDeps = {}): Database {
    const db = _getDbConnection(deps.config);

    db.useDatabase(deps.config.db.name);

    return db;
}

export const initDb = async (config: any) => {
    const db = _getDbConnection(config);

    const databases = await db.listDatabases();
    const dbExists = databases.reduce((exists, d) => exists || d === config.db.name, false);

    if (!dbExists) {
        await db.createDatabase(config.db.name);
    }

    db.close();
};
