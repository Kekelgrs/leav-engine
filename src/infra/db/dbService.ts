import {Database} from 'arangojs';

export default function(config: any) {
    const db = new Database({
        url: config.db.url
    });
    db.useDatabase(config.db.name);

    return db;
}
