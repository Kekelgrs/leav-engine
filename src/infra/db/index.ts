export {default} from './db';
export {default as dbService} from './dbService';
export {default as dbUtils} from './dbUtils';
