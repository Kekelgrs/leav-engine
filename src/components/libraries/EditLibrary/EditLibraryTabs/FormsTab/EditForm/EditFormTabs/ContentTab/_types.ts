import {DragObjectWithType} from 'react-dnd';
import {GET_ATTRIBUTES_attributes_list} from '../../../../../../../../_gqlTypes/GET_ATTRIBUTES';
import {FormElementTypes} from '../../../../../../../../_gqlTypes/globalTypes';
import {IKeyValue} from '../../../../../../../../_types/shared';

export const PLACEHOLDER_ID = '__placeholder__';

export enum UIElementTypes {
    DIVIDER = 'divider',
    FIELDS_CONTAINER = 'fields_container',
    TEXT_BLOCK = 'text_block',
    TABS = 'tabs'
}

export enum FieldTypes {
    TEXT_INPUT = 'input_field',
    DATE = 'date',
    CHECKBOX = 'checkbox',
    ENCRYPTED = 'encrypted',
    DROPDOWN = 'dropdown'
}

export enum DraggableElementTypes {
    RESERVE_LAYOUT_ELEMENT = 'RESERVE_LAYOUT_ELEMENT',
    ATTRIBUTE = 'ATTRIBUTE',
    FORM_ELEMENT = 'FORM_ELEMENT'
}

export interface IFormElement {
    id: string;
    order: number;
    type: FormElementTypes;
    uiElement: IUIElement;
    containerId: string;
    settings?: IKeyValue<any>;
    herited?: boolean;
}

export interface IUIElement {
    type: UIElementTypes | FieldTypes;
    component: JSX.Element;
    canDrop: (dropCandidate: IFormElement) => boolean;
    settings?: string[];
}

export interface IFormElementProps<T extends object> {
    elementData?: IFormElement;
    settings: T;
}

export interface ICommonFieldsSettings {
    label?: string;
    attribute?: GET_ATTRIBUTES_attributes_list;
}

export interface IFormElementPos {
    order: number;
    containerId: string;
}

export interface IFormBuilderDragObject<T extends IUIElement | IFormElement> extends DragObjectWithType {
    type: DraggableElementTypes;
    element: T;
    index: number;
    currentPos?: IFormElementPos;
    originPos?: IFormElementPos;
    dropAtPos?: IFormElementPos;
}
