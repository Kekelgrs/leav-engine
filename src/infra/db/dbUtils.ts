import {aql} from 'arangojs';
import {GeneratedAqlQuery} from 'arangojs/lib/cjs/aql-query';
import {CollectionType} from 'arangojs/lib/cjs/collection';
import {AwilixContainer} from 'awilix';
import {readdirSync} from 'fs';
import {IPluginsRepo} from 'infra/plugins/pluginsRepo';
import * as path from 'path';
import {isArray} from 'util';
import * as winston from 'winston';
import {IAttribute, IAttributeFilterOptions} from '_types/attribute';
import {IForm} from '_types/forms';
import {ILibrary, ILibraryFilterOptions} from '_types/library';
import {IList, IPaginationParams, ISortParams} from '_types/list';
import {IQueryInfos} from '_types/queryInfos';
import {IKeyValue} from '_types/shared';
import {ITree, ITreeFilterOptions} from '_types/tree';
import {IDbValueVersion, IValueVersion} from '_types/value';
import {collectionTypes, IDbService, IExecuteWithCount} from './dbService';
import runMigrationFiles from './helpers/runMigrationFiles';

export const MIGRATIONS_COLLECTION_NAME = 'core_db_migrations';

export interface IFindCoreEntityParams {
    collectionName: string;
    filters?: ITreeFilterOptions | ILibraryFilterOptions | IAttributeFilterOptions;
    strictFilters?: boolean;
    withCount?: boolean;
    pagination?: IPaginationParams;
    sort?: ISortParams;
    customFilterConditions?: IKeyValue<
        (filterKey: string, filterVal: string | boolean | string[], strictFilters: boolean) => GeneratedAqlQuery
    >;
    ctx: IQueryInfos;
}

export interface IDbUtils {
    migrate?(depsManager: AwilixContainer): Promise<void>;
    cleanup?(record: {}): any;
    convertToDoc?(obj: {}): any;
    isCollectionExists?(name: string): Promise<boolean>;
    findCoreEntity?<T extends ITree | ILibrary | IAttribute | IForm>(params: IFindCoreEntityParams): Promise<IList<T>>;
    convertValueVersionToDb?(version: IValueVersion): IDbValueVersion;
    convertValueVersionFromDb?(version: IDbValueVersion): IValueVersion;
    clearDatabase(): Promise<void>;
}

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.utils.logger'?: winston.Winston;
    'core.infra.plugins'?: IPluginsRepo;
    config?: any;
}

export default function({
    'core.infra.db.dbService': dbService = null,
    'core.utils.logger': logger = null,
    'core.infra.plugins': pluginsRepo = null,
    config = null
}: IDeps = {}): IDbUtils {
    /**
     * Create the collections used to managed db migrations
     * This can't be in a migration file because we do have to initialize it somewhere
     *
     */
    async function _initMigrationsCollection(): Promise<void> {
        const collections = await dbService.db.listCollections();

        const colExists = collections.reduce((exists, c) => exists || c.name === MIGRATIONS_COLLECTION_NAME, false);
        if (!colExists) {
            const collection = dbService.db.collection(MIGRATIONS_COLLECTION_NAME);
            await collection.create();
        }
    }

    /**
     * Return the filter's conditions based on key and val supplied.
     *
     * @param filterKey
     * @param filterVal
     * @param bindVars
     * @param index
     * @param strictFilters
     */
    function _getFilterCondition(
        filterKey: string,
        filterVal: string | boolean | string[],
        strictFilters: boolean
    ): GeneratedAqlQuery {
        const queryParts = [];

        // If value is an array (types or formats for example),
        // we call this function recursively on array and join filters with an OR
        if (Array.isArray(filterVal)) {
            if (filterVal.length) {
                const valParts = filterVal.map(val => _getFilterCondition(filterKey, val, strictFilters));
                queryParts.push(aql.join(valParts, ' OR '));
            }
        } else {
            if (filterKey === 'label') {
                // Search for label in any language
                const valParts = config.lang.available.map(l => aql`LIKE(el.label.${l}, ${filterVal}, true)`);
                queryParts.push(aql.join(valParts, ' OR '));
            } else {
                // Filter with a "like" on ID or exact value in other fields
                queryParts.push(
                    filterKey === '_key' && !strictFilters
                        ? aql`LIKE(el.${filterKey}, ${filterVal}, true)`
                        : aql`el.${filterKey} == ${filterVal}`
                );
            }
        }

        return aql.join(queryParts);
    }

    const ret = {
        /**
         * Run database migrations.
         * It takes all files present in migrations folder and run it if it's never been executed before
         *
         * @param depsManager
         */
        async migrate(depsManager: AwilixContainer): Promise<void> {
            await _initMigrationsCollection();
            const ctx: IQueryInfos = {
                userId: '0',
                queryId: 'run-migrations'
            };
            // Load already ran migrations
            const executedMigrations = await dbService.execute({
                query: `
                    FOR m IN core_db_migrations
                    RETURN m.file
                `,
                ctx
            });

            const _runMigrationFiles = (files, folder, prefix = null) =>
                runMigrationFiles({
                    files,
                    executedMigrations,
                    migrationsDir: folder,
                    prefix,
                    deps: {depsManager, dbService, logger},
                    ctx
                });

            /*** Core migrations ***/
            // Load migrations files
            const migrationsDir = path.resolve(__dirname, 'migrations');
            const migrationFiles = readdirSync(migrationsDir).filter(file => file.indexOf('.map') === -1);

            await _runMigrationFiles(migrationFiles, migrationsDir);

            /*** Plugins migrations ***/
            const plugins = pluginsRepo.getRegisteredPlugins();
            for (const plugin of plugins) {
                const pluginMigrationFolder = path.resolve(plugin.path + '/infra/db/migrations');
                const pluginMigrationFiles = readdirSync(pluginMigrationFolder).filter(
                    file => file.indexOf('.map') === -1
                );

                await _runMigrationFiles(pluginMigrationFiles, pluginMigrationFolder, plugin.infos.name);
            }
        },

        /**
         * Cleanup every system keys from an object coming from database.
         * _key is kept under 'id'
         *
         * @param obj
         * @return any   Cleaned up object
         */
        cleanup(obj: any): any {
            if (obj === null) {
                return null;
            }

            return Object.keys(obj).reduce((newObj: any, key) => {
                if (key === '_key') {
                    newObj.id = obj[key];
                } else if (key[0] !== '_') {
                    newObj[key] = obj[key];
                }

                return newObj;
            }, {});
        },

        /**
         * Convert an object to an object looking like a DB document
         * id is replaced by _key
         *
         * @param obj
         * @return any   DB document compatible object
         */
        convertToDoc(obj: {}): any {
            const newObj: any = {...obj};

            if (typeof newObj.id !== 'undefined') {
                newObj._key = newObj.id;
            }
            delete newObj.id;

            return newObj;
        },
        /**
         * Search core entities (libraries, attributes, trees)
         *
         * @param collectionName
         * @param filters
         * @param strictFilters
         */
        async findCoreEntity<T extends ITree | ILibrary | IAttribute>(
            params: IFindCoreEntityParams
        ): Promise<IList<T>> {
            const {
                collectionName = null,
                filters = null,
                strictFilters = false,
                withCount = false,
                pagination = null,
                sort = null,
                customFilterConditions = {},
                ctx = {}
            } = params;

            const collec = dbService.db.collection(collectionName);
            const queryParts = [aql`FOR el IN ${collec}`];

            if (filters !== null) {
                const dbFilters = ret.convertToDoc(filters);
                const filtersKeys = Object.keys(dbFilters);

                for (const filterKey of filtersKeys) {
                    const filterVal = dbFilters[filterKey];

                    // Caller can define some custom functions to generate filter condition (like looking on edges to
                    // filter on libraries linked to an attribute). So, if a custom function is define for this filter,
                    // we use it, otherwise we use the standard filters
                    const filterCondsFunc =
                        typeof customFilterConditions[filterKey] !== 'undefined'
                            ? customFilterConditions[filterKey]
                            : _getFilterCondition;
                    const conds = filterCondsFunc(filterKey, filterVal, strictFilters);

                    if (conds.query) {
                        queryParts.push(aql`FILTER`, conds);
                    }
                }
            }

            if (!!sort) {
                const field = sort.field === 'id' ? '_key' : sort.field;
                queryParts.push(aql`SORT el.${field} ${sort.order}`);
            }

            if (!!pagination) {
                queryParts.push(aql`LIMIT ${pagination.offset || 0}, ${pagination.limit}`);
            }

            queryParts.push(aql`RETURN el`);

            const query = aql.join(queryParts);
            const res = await dbService.execute<IExecuteWithCount | any[]>({query, withTotalCount: withCount, ctx});

            const results = !isArray(res) ? res.results : res;

            return {
                totalCount: withCount ? (res as IExecuteWithCount).totalCount : null,
                list: results.map(ret.cleanup)
            };
        },
        convertValueVersionToDb(version: IValueVersion): IDbValueVersion {
            return Object.keys(version).reduce((allVers, treeName) => {
                const {library, id} = version[treeName];

                allVers[treeName] = `${library}/${id}`;

                return allVers;
            }, {});
        },
        convertValueVersionFromDb(version: IDbValueVersion): IValueVersion {
            return version
                ? Object.keys(version).reduce((allVers, treeName) => {
                      const [library, id] = version[treeName].split('/');

                      allVers[treeName] = {
                          library,
                          id
                      };

                      return allVers;
                  }, {})
                : null;
        },
        async clearDatabase(): Promise<void> {
            // Drop all collections
            const cols = await dbService.db.listCollections();
            for (const col of cols) {
                const colType =
                    col.type === CollectionType.DOCUMENT_COLLECTION ? collectionTypes.DOCUMENT : collectionTypes.EDGE;
                await dbService.dropCollection(col.name, colType);
            }
        }
    };

    return ret;
}
