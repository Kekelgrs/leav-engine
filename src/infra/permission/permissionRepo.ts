import {aql} from 'arangojs';
import {IPermission, PermissionTypes, IPermissionsTreeTarget} from '../../_types/permissions';
import {IDbService} from '../db/dbService';
import {IDbUtils} from '../db/dbUtils';

export interface IPermissionRepo {
    savePermission(permData: IPermission): Promise<IPermission>;
    getPermissions(
        type: PermissionTypes,
        applyTo: string,
        usersGroupId: number,
        permissionTreeTarget?: IPermissionsTreeTarget
    ): Promise<IPermission | null>;
}

const PERM_COLLECTION_NAME = 'core_permissions';

export default function(dbService: IDbService, dbUtils: IDbUtils = null): IPermissionRepo {
    return {
        async savePermission(permData: IPermission): Promise<IPermission> {
            // Upsert in permissions collection
            const col = dbService.db.collection(PERM_COLLECTION_NAME);
            const searchObj = {
                type: permData.type,
                applyTo: permData.applyTo,
                usersGroup: permData.usersGroup,
                permissionTreeTarget: permData.permissionTreeTarget
            };

            const res = await dbService.execute(aql`
                UPSERT ${searchObj}
                INSERT ${permData}
                UPDATE ${permData}
                IN ${col}
                RETURN NEW
            `);

            return dbUtils.cleanup(res.pop());
        },
        async getPermissions(
            type: PermissionTypes,
            applyTo: string,
            usersGroupId: number,
            permissionTreeTarget: IPermissionsTreeTarget = null
        ): Promise<IPermission | null> {
            const col = dbService.db.collection(PERM_COLLECTION_NAME);
            const query = aql`
                FOR p IN ${col}
                FILTER p.type == ${type}
                    AND p.applyTo == ${applyTo}
                    AND p.usersGroup == ${usersGroupId}
                    AND p.permissionTreeTarget == ${permissionTreeTarget}
                RETURN p
            `;

            const res = await dbService.execute(query);

            return res.length ? res[0] : null;
        }
    };
}
