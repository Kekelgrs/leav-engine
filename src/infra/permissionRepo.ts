import {aql} from 'arangojs';
import {IPermission, PermissionTypes, IPermissionsTreeTarget} from '../_types/permissions';
import {IDbService} from './db/dbService';
import {IDbUtils} from './db/dbUtils';

export interface IPermissionRepo {
    savePermission(permData: IPermission): Promise<IPermission>;
    getPermissions(
        type: PermissionTypes,
        usersGroup: string,
        target?: IPermissionsTreeTarget
    ): Promise<IPermission | null>;
}

const PERM_COLLECTION_NAME = 'core_permissions';

export default function(dbService: IDbService, dbUtils: IDbUtils = null): IPermissionRepo {
    return {
        async savePermission(permData: IPermission): Promise<IPermission> {
            // Upsert in permissions collection
            const col = dbService.db.collection(PERM_COLLECTION_NAME);
            const searchObj = {
                type: permData.type,
                usersGroup: permData.usersGroup,
                target: permData.target
            };

            const res = await dbService.execute(aql`
                UPSERT ${searchObj}
                INSERT ${permData}
                UPDATE ${permData}
                IN ${col}
                RETURN NEW
            `);

            return dbUtils.cleanup(res.pop());
        },
        async getPermissions(
            type: PermissionTypes,
            usersGroup: string,
            target: IPermissionsTreeTarget = null
        ): Promise<IPermission | null> {
            const col = dbService.db.collection(PERM_COLLECTION_NAME);
            const query = aql`
                FOR p IN ${col}
                FILTER p.type == ${type}
                    AND p.usersGroup == ${usersGroup}
                    AND p.target == ${target}
                RETURN p
            `;

            const res = await dbService.execute(query);

            return res.length ? res[0] : null;
        }
    };
}
