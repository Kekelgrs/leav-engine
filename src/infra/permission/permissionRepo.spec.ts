import {Database} from 'arangojs';
import {IPermission, PermissionTypes, RecordPermissionsActions} from '../../_types/permissions';
import {IDbUtils} from '../db/dbUtils';
import permissionRepo from './permissionRepo';

describe('PermissionRepo', () => {
    describe('SavePermission', () => {
        test('Should save permission', async () => {
            const permData = {
                type: PermissionTypes.RECORD,
                applyTo: 'test_lib',
                usersGroup: '12345',
                actions: {
                    [RecordPermissionsActions.ACCESS]: true,
                    [RecordPermissionsActions.EDIT]: false,
                    [RecordPermissionsActions.DELETE]: false
                },
                permissionTreeTarget: {
                    id: 'test_lib/123445',
                    tree: 'test_tree'
                }
            };

            const permDataClean: IPermission = {
                type: PermissionTypes.RECORD,
                applyTo: 'test_lib',
                usersGroup: '12345',
                actions: {
                    [RecordPermissionsActions.ACCESS]: true,
                    [RecordPermissionsActions.EDIT]: false,
                    [RecordPermissionsActions.DELETE]: false
                },
                permissionTreeTarget: {
                    id: 123445,
                    library: 'test_lib',
                    tree: 'test_tree'
                }
            };

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([permData])
            };

            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: jest.fn().mockReturnValue(permDataClean)
            };

            const permRepo = permissionRepo(mockDbServ, mockDbUtils as IDbUtils);

            const savedPerm = await permRepo.savePermission(permDataClean);

            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatch(/UPSERT/);
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].bindVars).toMatchSnapshot();

            expect(savedPerm).toMatchObject(permDataClean);
        });
    });
    describe('GetPermissions', () => {
        test('Should return a permission', async () => {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([
                    {
                        actions: {
                            ACCESS: true,
                            EDIT: true
                        },
                        permissionTreeTarget: {
                            id: '123',
                            library: 'category',
                            tree: 'categories'
                        },
                        type: 'RECORD',
                        applyTo: 'test_lib',
                        usersGroup: '12345'
                    }
                ])
            };
            const permRepo = permissionRepo(mockDbServ);

            const perm = await permRepo.getPermissions(PermissionTypes.RECORD, 'test_lib', 12345, {
                id: '123',
                library: 'category',
                tree: 'categories'
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatch(/(?!FILTER)/);
            expect(mockDbServ.execute.mock.calls[0][0].query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].bindVars).toMatchSnapshot();
        });

        test('Should return null if no permissions', async () => {
            const mockDbServ = {db: new Database(), execute: global.__mockPromise([])};
            const permRepo = permissionRepo(mockDbServ);

            const perm = await permRepo.getPermissions(PermissionTypes.RECORD, 'test_lib', 12345, {
                id: '123',
                library: 'category',
                tree: 'categories'
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(perm).toBe(null);
        });
    });
});
