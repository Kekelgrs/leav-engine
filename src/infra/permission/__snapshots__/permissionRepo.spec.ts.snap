// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermissionRepo GetPermissions Should return a permission 1`] = `
"
                FOR p IN @@value0
                FILTER p.type == @value1
                    AND p.applyTo == @value2
                    AND p.usersGroup == @value3
                    AND p.permissionTreeTarget == @value4
                RETURN p
            "
`;

exports[`PermissionRepo GetPermissions Should return a permission 2`] = `
Object {
  "@value0": "core_permissions",
  "value1": "record",
  "value2": "test_lib",
  "value3": 12345,
  "value4": Object {
    "id": "123",
    "library": "category",
    "tree": "categories",
  },
}
`;

exports[`PermissionRepo SavePermission Should save permission 1`] = `
"
                UPSERT @value0
                INSERT @value1
                UPDATE @value1
                IN @@value2
                RETURN NEW
            "
`;

exports[`PermissionRepo SavePermission Should save permission 2`] = `
Object {
  "@value2": "core_permissions",
  "value0": Object {
    "applyTo": "test_lib",
    "permissionTreeTarget": Object {
      "id": 123445,
      "library": "test_lib",
      "tree": "test_tree",
    },
    "type": "record",
    "usersGroup": "users/12345",
  },
  "value1": Object {
    "actions": Object {
      "access": true,
      "delete": false,
      "edit": false,
    },
    "applyTo": "test_lib",
    "permissionTreeTarget": Object {
      "id": 123445,
      "library": "test_lib",
      "tree": "test_tree",
    },
    "type": "record",
    "usersGroup": "users/12345",
  },
}
`;
